File: backend/src/main/java/com/sudimango/MyStudyPal/dto/QuizAttemptDto.java
package com.sudimango.MyStudyPal.dto;

import java.time.Instant;
import java.util.List;
import java.util.Optional;

import com.sudimango.MyStudyPal.dto.QuizQuestionDto.QuizQuestionResponse;
import com.sudimango.MyStudyPal.entity.QuizAttempt;
import com.sudimango.MyStudyPal.entity.QuizAttemptAnswer;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;

// @formatter:off
public class QuizAttemptDto {

    public record AnswerSubmission(
        @NotBlank String questionId,
        List<String> userAnswer
    ) {}

    public record QuizAttemptAnswerResponse (
        String answerId,
        List<String> userAnswer,
        Boolean isCorrect,
        double pointsEarned,
        QuizQuestionResponse question
    ) {
        public QuizAttemptAnswerResponse(QuizAttemptAnswer answer) {
            this(
                answer.getAnswerId(),
                answer.getUserAnswer(),
                answer.getIsCorrect(),
                answer.getPointsEarned(),
                Optional.ofNullable(answer.getQuizAttemptQuestion())
                    .map(QuizQuestionResponse::new)
                    .orElse(null)
            );
        }
    }

    /**
     * Request
     */

    public record CreateQuizAttemptRequest(
        @NotNull Integer timeSpentSeconds,
        List<AnswerSubmission> answers
    ) {}

    /**
     * Response
     */

    public record CreateQuizAttemptResponse(
        String attemptId
    ) {}

    public record QuizAttemptResponse(
        String attemptId,
        double score,
        double maxScore,
        Instant startedAt,
        Instant completedAt
    ) {
        public QuizAttemptResponse(QuizAttempt attempt) {
            this(
                attempt.getAttemptId(),
                attempt.getScore(),
                attempt.getMaxScore(),
                attempt.getStartedAt(),
                attempt.getCompletedAt()
            );
        }
    }

    public record ShortAnswerGradingResponse(
        Double score
    ) {
        public ShortAnswerGradingResponse() {
            this(
                0.0
            );
        }
    }


}

File: frontend/src/lib/dto/quiz-attempt-dto.ts
import { QuizQuestionResponse } from "./quiz-question-dto";

export interface AnswerSubmission {
    questionId: string;
    userAnswer: string[];
}

export interface QuizAttemptAnswerResponse {
    answerId: string;
    userAnswer: string[];
    isCorrect: boolean;
    pointsEarned: number;
    question: QuizQuestionResponse;
}

/**
 * Request
 */

export interface CreateQuizAttemptRequest {
    timeSpentSeconds: number;
    answers: AnswerSubmission[];
}

/**
 * Response
 */

export interface CreateQuizAttemptResponse {
    attemptId: string;
}

export interface QuizAttemptResponse {
    attemptId: string;
    score: number;
    maxScore: number;
    startedAt: string;
    completedAt: string;
}
